File: Ex 1/ex1.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>


int i = 0;

// Note the return type: void*
void* adder(){
    for(int x = 0; x < 1000000; x++){
        i++;
    }
    return NULL;
}

void* subtract(){
    for(int x = 0; x < 1000000; x++){
        i--;
    }
    return NULL;
}


int main(){
    pthread_t adder_thr;
    pthread_create(&adder_thr, NULL, adder, NULL);
    pthread_t subtract_thr;
    pthread_create(&subtract_thr, NULL, subtract, NULL);

    pthread_join(subtract_thr, NULL);
    pthread_join(adder_thr, NULL);
    printf("Done: %i\n", i);
    return 0;
    
}

File: Ex 1/ex1.py
#!/usr/bin/python
from threading import Thread

i = 0

def adder():
    global i
    for x in range(0, 1000000):     
        i += 1

def subtract():
    global i
    for x in range(0, 1000000):     
        i -= 1


def main():
    adder_thr = Thread(target = adder)
    subtract_thr = Thread(target = subtract)

    adder_thr.start()
    subtract_thr.start()
    adder_thr.join()
    subtract_thr.join()
    print("Done: " + str(i))


main()
